26_Autoexposure/main.cpp

```diff
@@ -0,0 +1,874 @@
+			cmdbuf->pushConstants(m_lumaMeterPipeline->getLayout(), IShader::E_SHADER_STAGE::ESS_COMPUTE, 0, sizeof(pc), &pc);
+			cmdbuf->dispatch(1 + (SampleCount[0] - 1) / SubgroupSize, 1 + (SampleCount[1] - 1) / SubgroupSize);
+			cmdbuf->end();
+
+			{
+				IQueue::SSubmitInfo submit_infos[1];
+				IQueue::SSubmitInfo::SCommandBufferInfo cmdBufs[] = {
+					{
+						.cmdbuf = cmdbuf
+					}
+				};
+				submit_infos[0].commandBuffers = cmdBufs;
+				IQueue::SSubmitInfo::SSemaphoreInfo signals[] = {
+					{
+						.semaphore = m_lumaMeterSemaphore.get(),
+						.value = m_submitIx + 1,
+						.stageMask = asset::PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT
+					}
+				};
+				submit_infos[0].signalSemaphores = signals;
+
+				queue->submit(submit_infos);
+				queue->endCapture();
+			}
+
+			const ISemaphore::SWaitInfo wait_infos[] = {
+				{
+					.semaphore = m_lumaMeterSemaphore.get(),
+					.value = m_submitIx + 1
+				}
+			};
+			m_device->blockForSemaphores(wait_infos);
+		}
+
+		// Get EV
+		{
+			const auto memory_range = ILogicalDevice::MappedMemoryRange(
+				m_lumaGatherAllocation.memory.get(),
+				0ull,
+				m_lumaGatherAllocation.memory->getAllocationSize()
+			);
+
+			if (!m_lumaGatherAllocation.memory->getMemoryPropertyFlags().hasFlags(IDeviceMemoryAllocation::EMPF_HOST_COHERENT_BIT))
+				m_device->invalidateMappedMemoryRanges(1, &memory_range);
+
+			const uint32_t* buffData = reinterpret_cast<const uint32_t*>(m_lumaGatherAllocation.memory->getMappedPointer());
+
+			assert(m_lumaGatherAllocation.offset == 0); // simpler than writing out all the pointer arithmetic
+
+			m_EV = 0.0f;
+			for (int index = 0; index < SubgroupSize; index++) {
+				m_EV += static_cast<float>(buffData[index]) / (log2(LumaMinMax[1]) - log2(LumaMinMax[0])) + log2(LumaMinMax[0]);
+			}
+			m_EV /= (SampleCount[0] * SampleCount[1]);
+		}
+
+		// Render to swapchain
+		{
+			// Acquire
+			auto acquire = m_surface->acquireNextImage();
+			if (!acquire)
+				return;
+
+			auto queue = getGraphicsQueue();
+			auto cmdbuf = m_graphicsCmdBufs[0].get();
+			cmdbuf->reset(IGPUCommandBuffer::RESET_FLAGS::NONE);
+			auto ds = m_lumaPresentDS[1].get();
+
+			auto pc = AutoexposurePushData
+			{
+				.meteringWindowScaleX = MeteringWindowScale[0] * m_gpuImg->getCreationParameters().extent.width,
+				.meteringWindowScaleY = MeteringWindowScale[1] * m_gpuImg->getCreationParameters().extent.height,
+				.meteringWindowOffsetX = MeteringWindowOffset[0] * m_gpuImg->getCreationParameters().extent.width,
+				.meteringWindowOffsetY = MeteringWindowOffset[1] * m_gpuImg->getCreationParameters().extent.height,
+				.lumaMin = LumaMinMax[0],
+				.lumaMax = LumaMinMax[1],
+				.EV = m_EV,
+				.sampleCountX = SampleCount[0],
+				.sampleCountY = SampleCount[1],
+				.viewportSizeX = m_gpuImg->getCreationParameters().extent.width,
+				.viewportSizeY = m_gpuImg->getCreationParameters().extent.height,
+				.lumaMeterBDA = m_lumaGatherBDA
```

Comment:
if you're overdeclaring push constants to share the layout between two pipelines, then only build them once